src/render_tasks/d3d12_rtao_task.hpp

```diff
@@ -0,0 +1,297 @@
+				data.in_gbuffers = std::move(as_build_data.out_allocator->Allocate());
+				data.in_depthbuffer = std::move(as_build_data.out_allocator->Allocate());
+			}
+
+			// Bind output texture
+
+			d3d12::DescHeapCPUHandle rtv_handle = data.out_uav_from_rtv.GetDescriptorHandle();
+			d3d12::CreateUAVFromSpecificRTV(n_render_target, rtv_handle, 0, n_render_target->m_create_info.m_rtv_formats[0]);
+
+
+			// Bind g-buffers (albedo, normal, depth)
+			d3d12::DescHeapCPUHandle gbuffers_handle = data.in_gbuffers.GetDescriptorHandle();
+			d3d12::DescHeapCPUHandle depth_buffer_handle = data.in_depthbuffer.GetDescriptorHandle();
+
+			//cpu_handle = d3d12::GetCPUHandle(as_build_data.out_rt_heap, frame_idx, COMPILATION_EVAL(rs_layout::GetHeapLoc(params::rt_hybrid, params::RTHybridE::GBUFFERS)));
+
+			auto deferred_main_rt = data.in_deferred_main_rt = static_cast<d3d12::RenderTarget*>(fg.GetPredecessorRenderTarget<DeferredMainTaskData>());
+			d3d12::CreateSRVFromSpecificRTV(deferred_main_rt, gbuffers_handle, 1, deferred_main_rt->m_create_info.m_rtv_formats.data()[1]);
+			d3d12::CreateSRVFromDSV(deferred_main_rt, depth_buffer_handle);
+
+			if (!resize)
+			{
+				// Camera constant buffer
+				data.in_cb_camera_handle = static_cast<D3D12ConstantBufferHandle*>(n_render_system.m_raytracing_cb_pool->Create(sizeof(temp::RTHybridCamera_CBData)));
+
+				// Pipeline State Object
+				auto& rt_registry = RTPipelineRegistry::Get();
+				data.in_state_object = static_cast<D3D12StateObject*>(rt_registry.Find(state_objects::rt_ao_state_opbject))->m_native;
+
+				// Root Signature
+				auto& rs_registry = RootSignatureRegistry::Get();
+				data.in_root_signature = static_cast<D3D12RootSignature*>(rs_registry.Find(root_signatures::rt_ao_global))->m_native;
+			}
+
+			// Create Shader Tables
+			CreateShaderTables(device, data, 0);
+			CreateShaderTables(device, data, 1);
+			CreateShaderTables(device, data, 2);
+		}
+
+		inline void ExecuteAOTask(RenderSystem & render_system, FrameGraph & fg, SceneGraph & scene_graph, RenderTaskHandle & handle)
+		{
+			// Initialize variables
+			auto& n_render_system = static_cast<D3D12RenderSystem&>(render_system);
+			auto window = n_render_system.m_window.value();
+			auto device = n_render_system.m_device;
+			auto cmd_list = fg.GetCommandList<d3d12::CommandList>(handle);
+			auto& data = fg.GetData<RTAOData>(handle);
+			auto& as_build_data = fg.GetPredecessorData<wr::ASBuildData>();
+			fg.GetPredecessorData<wr::RTHybridData>(); //Wait for RTHybrid to avoid hyper threading issues.
+
+			d3d12::CreateOrUpdateTLAS(device, cmd_list, data.tlas_requires_init, data.out_tlas, as_build_data.out_blas_list);
+
+			// Wait for AS to be built
+			{
+				auto barrier = CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native);
+				cmd_list->m_native->ResourceBarrier(1, &barrier);
+			}
+
+			// Wait for AS to be built
+			cmd_list->m_native->ResourceBarrier(1, &CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native));
```

Comment:
double resource barrier

Reply:
This is indeed duplicate, though when looking at it I found something interesting. This code was inspired by the hybrid render task which is as follows: 
```cpp
// Rebuild acceleration structure a 2e time for fallback
			if (d3d12::GetRaytracingType(device) == RaytracingType::FALLBACK)
			{
				d3d12::CreateOrUpdateTLAS(device, cmd_list, data.tlas_requires_init, data.out_tlas, as_build_data.out_blas_list);

				auto barrier = CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native);
				cmd_list->m_native->ResourceBarrier(1, &barrier);
			}```
Yet fallback is not supported at all in RTAO, therefore I'd like to remove the resource barier entirely and move the fallback check from the demo framegraph to inside the AddAOTask (which should be renamed to AddRTAOTask while i think about it).

@VZout Will this implementation I described be correct or am I missing something?